activation/e2e/checkpoint_test.go

```diff
@@ -0,0 +1,255 @@
+		timesync.WithGenesisTime(genesis),
+		timesync.WithLayerDuration(layerDuration),
+		timesync.WithTickInterval(100*time.Millisecond),
+		timesync.WithLogger(zap.NewNop()),
+	)
+	require.NoError(t, err)
+	t.Cleanup(clock.Close)
+
+	localDB := localsql.InMemory()
+	nb, err := activation.NewNIPostBuilder(
+		localDB,
+		svc,
+		logger.Named("nipostBuilder"),
+		poetCfg,
+		clock,
+		validator,
+		activation.WithPoetServices(poetService),
+	)
+	require.NoError(t, err)
+
+	atxdata := atxsdata.New()
+	atxVersions := activation.AtxVersions{0: types.AtxV2}
+	edVerifier := signing.NewEdVerifier()
+	mpub := mocks.NewMockPublisher(ctrl)
+	mFetch := smocks.NewMockFetcher(ctrl)
+	mBeacon := activation.NewMockAtxReceiver(ctrl)
+	mTortoise := smocks.NewMockTortoise(ctrl)
+
+	atxHdlr := activation.NewHandler(
+		"local",
+		cdb,
+		atxdata,
+		edVerifier,
+		clock,
+		mpub,
+		mFetch,
+		goldenATX,
+		validator,
+		mBeacon,
+		mTortoise,
+		logger,
+		activation.WithAtxVersions(atxVersions),
+	)
+
+	tab := activation.NewBuilder(
+		activation.Config{GoldenATXID: goldenATX},
+		db,
+		atxdata,
+		localDB,
+		mpub,
+		nb,
+		clock,
+		syncer,
+		logger,
+		activation.WithPoetConfig(poetCfg),
+		activation.WithValidator(validator),
+		activation.BuilderAtxVersions(atxVersions),
+	)
+	tab.Register(sig)
+
+	var atx0ID types.ATXID
+	mpub.EXPECT().Publish(gomock.Any(), pubsub.AtxProtocol, gomock.Any()).DoAndReturn(
+		func(ctx context.Context, p string, msg []byte) error {
+			var watx wire.ActivationTxV2
+			codec.MustDecode(msg, &watx)
+			atx0ID = watx.ID()
+			peer := peer.ID(p)
+			mFetch.EXPECT().RegisterPeerHashes(peer, gomock.Any())
+			mFetch.EXPECT().GetPoetProof(gomock.Any(), gomock.Any())
+			mBeacon.EXPECT().OnAtx(gomock.Any())
+			mTortoise.EXPECT().OnAtx(gomock.Any(), gomock.Any(), gomock.Any())
+			err := atxHdlr.HandleGossipAtx(ctx, peer, msg)
+			require.NoError(t, err)
+			return err
+		},
+	)
+
+	require.NoError(t, tab.BuildInitialPost(ctx, sig.NodeID()))
+	require.NoError(t, tab.PublishActivationTx(ctx, sig))
+
+	// Execute checpoint-recovery
+	// 1. Generate checkpoint
+	require.NoError(t, accounts.Update(db, &types.Account{}))
+	snapshot := clock.CurrentLayer()
+	fs := afero.NewMemMapFs()
+	dir, err := afero.TempDir(fs, "", "Generate")
+	require.NoError(t, err)
+	err = checkpoint.Generate(ctx, fs, db, dir, snapshot, 1)
+	require.NoError(t, err)
+
+	// 2. Recover from checkpoint
+	recoveryCfg := checkpoint.RecoverConfig{
+		GoldenAtx: goldenATX,
+		DataDir:   t.TempDir(),
+		NodeIDs:   []types.NodeID{sig.NodeID()},
+		Restore:   snapshot,
+	}
+	filename := checkpoint.SelfCheckpointFilename(dir, snapshot)
+
+	var newDB *sql.Database
+	createDb := func() (*sql.Database, error) {
+		newDB = sql.InMemory()
+		return newDB, nil
+	}
+
+	data, err := checkpoint.RecoverFromLocalFile(ctx, logger, db, localDB, fs, &recoveryCfg, filename, createDb)
+	require.NoError(t, err)
+	require.NotNil(t, newDB)
+	require.Nil(t, data)
```

Comment:
The `creadDb` hack isn't needed for this test:

```suggestion
	// 2. Recover from checkpoint
	recoveryCfg := checkpoint.RecoverConfig{
		GoldenAtx: goldenATX,
		DataDir:   t.TempDir(),
		NodeIDs:   []types.NodeID{sig.NodeID()},
		Restore:   snapshot,
	}
	filename := checkpoint.SelfCheckpointFilename(dir, snapshot)

	data, err := checkpoint.RecoverFromLocalFile(ctx, logger, db, localDB, fs, &recoveryCfg, filename)
	require.NoError(t, err)
	require.Nil(t, data)

	newDB, err := sql.Open("file:" + filepath.Join(recoveryCfg.DataDir, recoveryCfg.DbFile))
	require.NoError(t, err)
	defer newDB.Close()
```

`newDB` can now be passed to all services just like with the callback.